Refactor Sharing

OlegZharkov opened the pull request:
This PR is a tight collaboration between @OlegZharkov and @davelopez  

## What did you do? 
Major refactor of Sharing functionality (back-end and front-end) and additional Vuefication of `Share with Individual Users`.

### UI Modernization

- Re-implement `share with Individual Users` in Vue.
    * Vue-multiselect for Admin/expose_users (autocomplete reactive select)
    * Share history with private datasets, done with `b-list` and made it 'toggle-able'

- A few improvements  in existent code (stuff like simplification of clipboard copy and icons consistency)
- Drop `history/share.mako`.
- Removed string-based requests, adapted with new paths
- Update Seleniun tests

### Backend

- Replaced old UI controller endpoints with FastAPI endpoints for sharing resources (histories, visualizations and pages):

  - `GET /api/{resource}/{id}/sharing`: gets the current sharing status of the given item.
  - `PUT /api/{resource}/{id}/publish`: makes this item public and accessible by a URL link.

  - `PUT /api/{resource}/{id}/unpublish`: removes this item from the published list.
  - `PUT /api/{resource}/{id}/enable_link_access`: makes this item accessible by a URL link.

  - `PUT /api/{resource}/{id}/disable_link_access`: makes this item inaccessible by a URL link.

  - `PUT /api/{resource}/{id}/share_with_users`: shares this item with specific users. The list of users can be updated, a.k.a. removed users will no longer share the resource.

  - `PUT /api/{resource}/{id}/slug`: sets a new slug for this shared item.

- The common sharing logic is implemented by the `ShareableService` which makes use of the existing `SharableModelManager` and `SharableModelSerializer` implementations of each resource.

- The particular behavior for sharing histories that may contain private or innaccessible datasets has been simplified a bit and modelled by using the optional `ShareWithStatus.extra` field.

- Added *temporal* legacy API endpoints for compatibility until UWSGI support is dropped.

- Adapted the common API tests for sharing and added `SharingHistoryTestCase` class to tests the particularities of sharing histories containing private datasets.

## Why did you make this change?

Preliminary step for #11667 and modernization of the sharing UI and API.

## How to test the changes?

- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] Instructions for manual testing are as follows:
  1. Go to Histories.
  2. Choose a history and select `Share or Publish`.
  3. Share it with one or more specific users.
  4. The steps above can be done also with `Pages` and `Visualizations`.

## License

- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

## For UI Components
- [x] I've included a screenshot of the changes


**Admin:**

![663687bb-3552-46ee-ab8e-5e7973fb221b](https://user-images.githubusercontent.com/15801412/133628747-e64fec79-861f-44f6-bfe3-38b439fb20d4.gif)

**User/Share History:**
![Sharing](https://user-images.githubusercontent.com/15801412/133628846-7e9fe49c-aa23-43d0-9966-55957647c93f.gif)

OlegZharkov committed:
initial implementation

davelopez committed:
Refactor ShareableService

This is an attempt to simplify the sharing logic and make use of the existing SharableModelManager abstraction.

OlegZharkov committed:
adopt UI to new backend
make multi-select work

OlegZharkov committed:
add expose_user_email in config

OlegZharkov committed:
preliminary template for expose_user_email=False

davelopez committed:
Get users either by id or email

OlegZharkov committed:
initial implementation of expose_email=false

OlegZharkov committed:
fix saving id in users autocomplete

OlegZharkov committed:
check if email was already shared

davelopez committed:
Add new API endpoint to change slug

davelopez committed:
Use new slug API on UI

davelopez committed:
Add FastAPI sharing endpoints to Histories and Visualizations

davelopez committed:
Fix linting

davelopez committed:
Add temporal legacy endpoints for setting the slug

davelopez committed:
Add test for set_slug

OlegZharkov committed:
fix user_ids, rework share buttons, fix layout, fix warnings, add tooltip

OlegZharkov committed:
Fix admin user-autocomplete. delete SelectUsers component

davelopez committed:
Refactor sharing API to full rest + tests

davelopez committed:
Add temporal legacy endpoints for the new sharing API

OlegZharkov committed:
adapt client to the new endpoints

OlegZharkov committed:
re-work erros rendoring

OlegZharkov committed:
linting

OlegZharkov committed:
simplify copy to clipboard

davelopez committed:
Rename share and unshare actions

This is to emphasize that the action requires a target user or list of users

davelopez committed:
WIP deal with sharing histories with dataset restrictions

davelopez committed:
Handle sharing histories with restricted datasets

davelopez committed:
Add tests for sharing restricted histories

davelopez committed:
Send only unique datasets previews when sharing

davelopez committed:
Add error if there is no dataset than can be shared in a history

OlegZharkov committed:
implement private datasets in history usecase

OlegZharkov committed:
implement share option

davelopez committed:
Add more history sharing tests and move them to SharingHistoryTestCase

davelopez committed:
Add num_accessible datasets to ShareHistoryExtra and fix some conditions

OlegZharkov committed:
implement 'can_change available usecase'

OlegZharkov committed:
fix eslint wine about v-key

OlegZharkov committed:
fix btn and icon methods mismatch

OlegZharkov committed:
make share button disabled if the field is empty, change button style

OlegZharkov committed:
add cancel button

OlegZharkov committed:
fix admin interface

davelopez committed:
Do not send extra field when sharing with `no_changes` + update tests

OlegZharkov committed:
clear email field on succesfull share

OlegZharkov committed:
fix dismissible alert

OlegZharkov committed:
move tooltip to the bottom

davelopez committed:
Avoid sharing attempt if all users are invalid

OlegZharkov committed:
delete history share.mako

OlegZharkov committed:
don't allow duplicate errors

davelopez committed:
Remove legacy code

OlegZharkov committed:
add csv support

OlegZharkov committed:
fix `already shared with this user` error handling in csv usecase

OlegZharkov committed:
fix selenium test

davelopez committed:
Fix share condition with option

OlegZharkov committed:
fix remaining selenium

davelopez committed:
Avoid sharing with duplicated users in the same request

mvdbeek commented on client/src/utils/clipboard.js:
```diff
@@ -4,7 +4,7 @@ import { Toast } from "ui/toast";
     if (navigator.clipboard) {
         navigator.clipboard.writeText(text).then(() => {
-            Toast.info(notificationText);
+            if (notificationText) Toast.info(notificationText);
```
Should this be
```suggestion
            if (notificationText) {
                Toast.info(notificationText);
            }
```
? It works, but I don't think we use this commonly across the codebase, not sure we should start doing it now.

I'm also curious if we should even write to the clipboard if there is no text ?

OlegZharkov replied:
> I'm also curious if we should even write to the clipboard if there is no text ?

This is not the text that goes into clipboard. It's just the text that appears in a Toast, if someone doesn't want to see a Toast `notificationText` will be empty.

> It works, but I don't think we use this commonly across the codebase, not sure we should start doing it now.

The behavior of `if()` is defined in our eslint. The rules tell us, if it doesn't have curly brackets it should be one line. If you don't like it, it's ok, I will change it here. But in general, if we don't support it, we should change the config of eslint. It's in the code and I used it in many places already.

mvdbeek commented on lib/galaxy/managers/histories.py:
```diff
@@ -171,6 +212,68 @@ def non_ready_jobs(self, history):
+        extra = ShareHistoryExtra()
+        history = cast(model.History, item)
+        if history.empty:
+            errors.add("You cannot share an empty history.")
```
Why not ? Seems fine, for instance if you know you want to share everything that is eventually put into this history.

OlegZharkov replied:
Well, it wasn't really our choice. It was in the previous implementation
![image](https://user-images.githubusercontent.com/15801412/117003587-534d5f80-ace5-11eb-8e33-383b48e74fd3.png)


Do you want to change it?

mvdbeek replied:
We can leave it like that for now.

mvdbeek replied:
Isn't that what the failing selenium test is about ? 
```
>               raise AssertionError(message)
E               AssertionError: Text [You cannot share resources with yourself] expected inside of [×
E               You cannot share an empty history.] but not found.
```
I'm pretty sure the API supported sharing empty histories. It's a bit of an arbitrary restriction IMO.

davelopez replied:
We totally agree that avoiding sharing an empty history seems like an arbitrary restriction and looking at the selenium tests doesn't look like it ever uploaded data to the history (which may explain why they are failing now and not before), but then, we are confused with the current behavior in the dev branch if you try to share an empty history with other users.

**Edit:** I think I may know what has changed to make these tests fail now. Looking again at the legacy code:
https://github.com/galaxyproject/galaxy/blob/386ab809fa46acbb4c8d62757c40f3515d5bfca5/lib/galaxy/webapps/galaxy/controllers/history.py#L786
The previous error handling was just returning a string with a concatenation of errors, and the `You cannot share an empty history.` seems to be returned **only** if there were no other errors... In the new implementation, we return a list containing all the errors encountered in the request, so that will explain why those selenium tests were passing before, the error message about the empty history was not getting to the response as there were other errors (the ones that were being tested).

mvdbeek commented on lib/galaxy/managers/histories.py:
```diff
@@ -28,6 +38,37 @@
+        title="Can Change",
+        description=(
+            "A collection of datasets that are not accessible by one or more of the target users "
+            "but, the user sharing the history, has permission to make them accessible for others."
```
```suggestion
            "A collection of datasets that are not accessible by one or more of the target users "
            "and that can be made accessible for others by the user sharing the history."
```

mvdbeek commented on lib/galaxy/managers/histories.py:
```diff
@@ -28,6 +38,37 @@
+        title="Cannot Change",
+        description=(
+            "A collection of datasets that are not accessible by one or more of the target users "
+            "and, the user sharing the history, does not have permission to make them accessible for others."
```
```suggestion
            "and that cannot be made accessible for others by the user sharing the history."
```

mvdbeek commented on lib/galaxy/managers/histories.py:
```diff
@@ -171,6 +212,68 @@ def non_ready_jobs(self, history):
+
+            user_roles = user.all_roles()
+            # Only deal with datasets that have not been purged
+            for hda in history.activatable_datasets:
```
It would be good to move this for loop into a query that returns "action tuples" like we do in https://github.com/mvdbeek/galaxy/blob/e57a04f3002229cf29d38e7837883a2f4019f2ff/lib/galaxy/model/__init__.py#L4239. These can then be used with https://github.com/mvdbeek/galaxy/blob/e57a04f3002229cf29d38e7837883a2f4019f2ff/lib/galaxy/model/security.py#L480 and a variant of https://github.com/mvdbeek/galaxy/blob/e57a04f3002229cf29d38e7837883a2f4019f2ff/lib/galaxy/model/security.py#L849 that updates the permissions in one statement.

The way we do this now is that we're firing N lazy loads per dataset, as well as N flushes in make_dataset_public or privately_share_dataset.

Okay, writing this down I realize that's some extra work we can do in another PR.

davelopez replied:
Yeah, this part felt a bit suboptimal while refactoring the legacy code, but didn't want (and didn't know exactly how) to change many parts.
I will open an issue to track this optimization and address it in a separate PR. Thanks for the tips!!

davelopez replied:
Tracked in #11926

mvdbeek commented on lib/galaxy/managers/histories.py:
```diff
@@ -171,6 +212,68 @@ def non_ready_jobs(self, history):
+                self.user_share_model.table.c.user_id == user.id,
+                self.user_share_model.table.c.history_id == history.id,
+            )
+        ).count() > 0
```
```suggestion
        ).exists()
```

mvdbeek commented on lib/galaxy/managers/sharable.py:
```diff
@@ -163,9 +283,12 @@ def share_with(self, item, user, flush=True):
+        # allow user to be a list and call recursively
         if isinstance(user, list):
-            return [self.share_with(item, _, flush=False) for _ in user]
+            user_assocs = [self.share_with(item, _, flush=False) for _ in user]
```
Just grepping around I don't think user is currently a list. Maybe drop this, we can add this when we really need it, but I doubt we will need it.

davelopez replied:
If the user who shares writes multiple emails separated by `,` in the UI, this can be an actual list of users:
https://github.com/OlegZharkov/galaxy/commit/0fa8470bd86be7db6f79560a9a30bbf1958b4880#diff-f0de4cf20668343a55f08ea7b70f0e138ee9cbe891e637ba359594422d54fafeR719

davelopez replied:
With the new approach of sharing and unsharing with a single list, this is no longer needed so the code has been removed in https://github.com/galaxyproject/galaxy/pull/11701/commits/508f96b65ae65c250ed77f07b13a09dc3620427c as suggested.
Thanks @mvdbeek!

mvdbeek commented on lib/galaxy/managers/sharable.py:
```diff
@@ -194,7 +317,10 @@ def unshare_with(self, item, user, flush=True):
+            user_assocs = [self.unshare_with(item, _, flush=False) for _ in user]
+            if flush:
+                self.session().flush()
+            return user_assocs
```
Same here, we can drop this I think

davelopez replied:
Same as above here, the API allows to provide multiple users to share and unshare in a single request, but we can limit it to one at a time if it makes sense.

davelopez replied:
See comment above https://github.com/galaxyproject/galaxy/pull/11701#discussion_r710048224

mvdbeek commented on lib/galaxy/managers/sharable.py:
```diff
@@ -596,97 +662,104 @@ class ShareableService:
+            raise exceptions.MessageException("Missing required user IDs or emails")
+        send_to_users: Set[User] = set()
+        send_to_err: Set[str] = set()
+        for string in set(emails_or_ids):
```
```suggestion
        for email_or_id in set(emails_or_ids):
```

mvdbeek commented on lib/galaxy/managers/sharable.py:
```diff
@@ -596,97 +662,104 @@ class ShareableService:
+                    send_to_user = None
+
+            if not send_to_user:
+                send_to_err.add(f"{string} is not a valid Galaxy user.")
```
We should probably not return this (or at least not for regular users), since that is a batch way to check which users are registered on a Galaxy instance. I would assume that is why we haven't done that before.

davelopez replied:
Totally agree here, we even discussed it when refactoring the code, but this message was already in the legacy code, that is why we maintained it here. I'm happy to replace it with something different, but how can we help users that get confused when they enter a wrong email and nothing happens?

Something similar happened recently, and even this `{email} is not a valid Galaxy user` wasn't of much help :)
https://help.galaxyproject.org/t/problem-with-getting-shared-histories/5886/3?u=davelopez

mvdbeek commented on lib/galaxy/managers/sharable.py:
```diff
@@ -596,97 +662,104 @@ class ShareableService:
+        send_to_err: Set[str] = set()
+        for string in set(emails_or_ids):
+            string = string.strip()
+            if not string:
```
```suggestion
            email_or_id = email_or_id.strip()
            if not email_or_id:
```

mvdbeek commented on lib/galaxy/managers/sharable.py:
```diff
@@ -596,97 +662,104 @@ class ShareableService:
+
+            send_to_user = None
+            if '@' in string:
+                email_address = string
```
```suggestion
            if '@' in email_or_id:
                email_address = email_or_id
```

mvdbeek commented on lib/galaxy/managers/sharable.py:
```diff
@@ -596,97 +662,104 @@ class ShareableService:
+                    filters=[User.table.c.deleted == false()])
+            else:
+                try:
+                    decoded_user_id = trans.security.decode_id(string)
```
```suggestion
                    decoded_user_id = trans.security.decode_id(email_or_id)
```

mvdbeek commented on lib/galaxy/managers/sharable.py:
```diff
@@ -596,97 +662,104 @@ class ShareableService:
+                    send_to_user = None
+
+            if not send_to_user:
+                send_to_err.add(f"{string} is not a valid Galaxy user.")
```
```suggestion
                send_to_err.add(f"{email_or_id} is not a valid Galaxy user.")
```

mvdbeek commented on client/src/components/Sharing/Sharing.vue:
```diff
@@ -0,0 +1,469 @@
+            type: String,
+            required: true,
+        },
+        model_class: {
```
The props should be camelCased

OlegZharkov replied:
are you sure? that's not really my code. Is it a rule of smth? Because we use `_` as a separator in almost every component across UI. I don't think, it's an issue, but it's ok, I will change it

dannon replied:
Yes, it should be camelCased.  Right now it's just a warning because there are so many of them that are broken, but this is a strongly recommended eslint rule we want to enable.

The reason they're still not all fixed is that when props are passed based on dict contents from the backend, where it can actually be snake_case that gets dictified into json, so each occasion requires examination.

(https://github.com/galaxyproject/galaxy/blob/dev/client/.eslintrc.js#L31)

mvdbeek commented on client/src/components/Sharing/Sharing.vue:
```diff
@@ -0,0 +1,469 @@
+    data() {
+        const Galaxy = getGalaxyInstance();
+        return {
+            isExposeEmail: Galaxy.config.expose_user_email || Galaxy.user.attributes.is_admin,
```
Can you use the ConfigProvider to access config values  and the CurrentUser provider to access the current user?

OlegZharkov replied:
I did it. but I have mixed feeling about it. Instead of one line solution, I had to define and register 2 different components. Is there any other way, where I can just call a function? Anyway, it's fine for now

dannon replied:
With Vue3's composition API I think we can make this a lot cleaner.  We absolutely do need to move away from ever touching `Galaxy` global object, though, so this (using the providers) is the right next step.

OlegZharkov replied:
is it possible to call it directly from Vuex without defining `mapGetters`?

mvdbeek commented on client/src/components/Sharing/Sharing.vue:
```diff
@@ -0,0 +1,469 @@
+            this.showUrl = false;
+        },
+        assignItem(newItem) {
+            if (newItem.errors) this.errors = newItem.errors;
```
Shouldn't this add individual items instead of replacing errors ?
Also see my comment below about oneline if statements.

OlegZharkov replied:
Well, I don't want to show old errors from the previous request. Right before this call I do `this.errors = [];`

> Also see my comment below about oneline if statements.

Once we decide the linting faith in https://github.com/galaxyproject/galaxy/pull/11925, sure. Once, we get it merged prettier will do one style for the whole application. So I'll wait for a bot to do it :) . I hope you're fine with this.

mvdbeek replied:
> Well, I don't want to show old errors from the previous request. Right before this call I do `this.errors = [];`
> 
> >

But that's confusing then, because you're not undoing the previous failed operation in the interface, or did I miss that ?

OlegZharkov replied:
> > Well, I don't want to show old errors from the previous request. Right before this call I do `this.errors = [];`
> > >
> 
> But that's confusing then, because you're not undoing the previous failed operation in the interface, or did I miss that ?

No worries, it's getting re-rendered here. Old errors are gone after a new request https://github.com/galaxyproject/galaxy/blob/a3516c8cfe3921916f32cdb30a027feb186ce713/client/src/components/Sharing/Sharing.vue#L5

mvdbeek commented on client/src/components/Sharing/Sharing.vue:
```diff
@@ -0,0 +1,469 @@
+                if (user_id.includes(",")) {
+                    user_ids = user_id.replace(/ /g, "").split(",");
+                } else user_ids = [user_id];
+            }
```
```suggestion
            const user_ids = user_id ? user_id.replace(/ /g, "").split(",") : undefined;
```
Neither space nor comma are allowed in user ids or emails, so we don't need the if/else.

OlegZharkov replied:
yeah, this one line solution is better, thanks :+1:

mvdbeek commented on client/src/components/Sharing/Sharing.vue:
```diff
@@ -0,0 +1,469 @@
+        onPublish(published) {
+            if (published) {
+                this.item.importable = true;
+                this.setSharing(this.actions.publish);
```
Shouldn't this be `this.setSharing(undefined, this.actions.publish);` (and everywhere else where you pass a single `share_option` argument)? My JS knowledge might be failing me here.

OlegZharkov replied:
Sorry, I didn't get you.
`setSharing(action, user_id, share_option) `
if you call `this.setSharing(undefined, this.actions.publish);`, then `action` would be `undefined` and `user_id` is `this.actions.publish`.

mvdbeek replied:
Ah, I think I missed that the first argument is `action`. Yeah, that makes sense.

mvdbeek commented on client/src/components/Sharing/Sharing.vue:
```diff
@@ -0,0 +1,469 @@
+                    this.errors = [];
+                    this.assignItem(data);
+                    if (data.extra && data.extra.can_share) this.shareWithEmail = "";
+                    else this.shareWithEmail = user_id || "";
```
I think this might be more readable as either a ternary statement or a regular if / else with curly brackets.

dannon replied:
I could swear there was another comment about this but I can't find it right now.  Anyway, xref #11925 as an option for standardizing this.

OlegZharkov replied:
I will wait for https://github.com/galaxyproject/galaxy/pull/11925 on this. I hope you're fine with this

mvdbeek commented on client/src/components/Sharing/Sharing.vue:
```diff
@@ -0,0 +1,469 @@
+                axios
+                    .get(`${getAppRoot()}api/users?f_email=${searchValue}`)
+                    .then((response) => {
+                        this.usersList = response.data;
```
You might want to restrict this to a minimum search value and possibly limit this on the backend. Returning all accounts containing `a` on the large instances is probably not what we'd want to do. 

We also probably want to make this conditional on `isExposeEmail`, if I am looking at this right the multiselect is not within `<div v-if="!isExposeEmail">`, so we should at least skip making the request here.

OlegZharkov replied:
> You might want to restrict this to a minimum search value and possibly limit this on the backend. Returning all accounts containing a on the large instances is probably not what we'd want to do.

done

> We also probably want to make this conditional on isExposeEmail, if I am looking at this right the multiselect is not within <div v-if="!isExposeEmail">, so we should at least skip making the request here.

it's conditional on `isExposeEmail`

mvdbeek reviewed (commented):
This is awesome, however there are still some issue to iron out.
<img width="556" alt="Screenshot 2021-05-04 at 12 24 48" src="https://user-images.githubusercontent.com/6804901/116990807-ec737a80-acd3-11eb-92a5-76ba54cf13d1.png">
may or may not be fixed by fixed by fixing the arguments being passed to setSharing.

I've also seen this while editing the slug of a page:
<img width="555" alt="Screenshot 2021-05-04 at 12 43 10" src="https://user-images.githubusercontent.com/6804901/116992487-6147b400-acd6-11eb-8894-a5dc13f85a41.png">


I think the styling of the select box doesn't fit the rest of Galaxy (unfortunately I have no suggestion on how to do it better except for looking at other select fields in Galaxy). It should probably be blue for highlighted items and a bit smaller.

I'm also missing an OK / Cancel button here. Just firing the request on select seems confusing to me, but I could be convinced otherwise.

OlegZharkov committed:
Update client/src/utils/clipboard.js

Co-authored-by: Marius van den Beek <m.vandenbeek@gmail.com>

davelopez committed:
Apply suggestions from code review

Co-authored-by: Marius van den Beek <m.vandenbeek@gmail.com>

OlegZharkov committed:
fix slug

OlegZharkov committed:
use camelcase

OlegZharkov committed:
use ConfigProvider and CurrentUser

OlegZharkov committed:
make import more stable

OlegZharkov committed:
oneline solution for user_ids

Co-authored-by: Marius van den Beek <m.vandenbeek@gmail.com>

davelopez committed:
Fix is_history_shared_with query + test multiple shares with same user

OlegZharkov committed:
change multiselect style

OlegZharkov committed:
add submit button, change style

OlegZharkov committed:
make it identical to bootsrap blue

OlegZharkov committed:
apply new linting rules

OlegZharkov committed:
require user to enter at least 3 letters to start email-autocompletion

davelopez committed:
Fix wrong number of accessible datasets

When the history was already shared with the user the `total_dataset_count` was not updated, leading to wrongly report that the history contained no accessible datasets.
This should be slightly faster now since we are retrieving only once the datasets.

davelopez committed:
Adapt tests to use JSON payload

See https://github.com/galaxyproject/galaxy/pull/12152

davelopez committed:
Fix Sharing component import

davelopez committed:
Merge share and unshare in the same request

Instead of having two different requests for adding or removing users from a shared resource now a simple list of users is sent in the request.
The resource will be shared with new users and unshared with users that are no more on that list.

davelopez committed:
Rename `set_sharing_with_users` to `share_with_users`

- Fix legacy routes in buildapp.py

OlegZharkov committed:
preliminary reworked interface

OlegZharkov committed:
fix the url path

davelopez committed:
Remove test assert no longer applicable

OlegZharkov committed:
notify user when the dataset is shared

OlegZharkov committed:
fix toast condition, fix placeholder and 3 char condition

davelopez committed:
Use unique resource name in tests

OlegZharkov committed:
rework style, fix a lot of things

OlegZharkov committed:
fix dataset warning, change buttons style

OlegZharkov committed:
add on focus out

OlegZharkov committed:
fix success alert condition

OlegZharkov committed:
remake condition

OlegZharkov committed:
fix cornercase, when only one dataset available, but user is able to make it public

OlegZharkov committed:
implement disable

davelopez committed:
Fix FastAPI route in visualizations

davelopez committed:
Remove unused code

This is efectively no longer used now so it can be removed as Marius suggested.

OlegZharkov committed:
fix user candidates array reference

OlegZharkov committed:
remove debug

OlegZharkov committed:
change id to emails on enter

OlegZharkov committed:
fix selenium test_sharing_valid_by_id

OlegZharkov committed:
fix HistorySharingTestCase::test_unsharing

mvdbeek commented on client/src/components/Sharing/Sharing.vue:
```diff
@@ -0,0 +1,593 @@
+        </div>
+        <br />
+        <div v-if="!hasUsername">
+            <div>To make a {{ modelClass }} accessible via link or publish it, you must create a public username:</div>
+            <form class="form-group" @submit.prevent="setUsername()">
+                <input class="form-control" type="text" v-model="newUsername" />
+            </form>
+            <b-button type="submit" variant="primary" @click="setUsername()">Set Username</b-button>
```
This should be its own component, and I think we probably want to do some additional validation around the username. I haven't checked what we do for the registration, but I think we list what makes a  valid username there?

Maybe we already have a component for setting the username you could reuse ?
Maybe you could just redirect the user to the user preferences if that's too much work ?

OlegZharkov replied:
To be honest I never encountered this particular interface. This code wasn't written by us. It shows green, because we just copied this file in another directory.
Anyway, can you please give a hint, how do you create user without username? :sweat_smile: 
register page doesn't allow this, is it supposed to be done with a script?

mvdbeek replied:
You can register a user via ldap, and then the user may not have a username. @nsoranzo @dannon do you remember the specifics here ? I vaguely remember we're now actually setting a username in all cases ?

If you want to test this out you can delete the username in the database for an existing user, I guess.

OlegZharkov replied:
Ok thanks, I managed to see it :+1: . I will refactor setUsername in a separate PR for that, if it's fine by you.

mvdbeek replied:
It is. I'll just run some tests now and probably merge later.

mvdbeek commented on client/src/components/Sharing/Sharing.vue:
```diff
@@ -0,0 +1,593 @@
+                        <b-col>
+                            <b-card
+                                border-variant="primary"
+                                header="How would like to proceed?"
```
```suggestion
                                header="How would you like to proceed?"
```

mvdbeek commented on client/src/components/Sharing/Sharing.vue:
```diff
@@ -0,0 +1,593 @@
+            elementsNotFoundWarning: "No elements found. Consider changing the search query.",
+            ready: false,
+            threeCharactersEntered: true,
+            hasUsername: Galaxy.user.get("username"),
```
We want to avoid accessing global Galaxy like this. I think it would be best to accept a user prop and move CurrentUser to an outer component. (Or to break this component into smaller pieces and use the CurrentUser provider as appropriate)

mvdbeek commented on client/src/components/Sharing/Sharing.vue:
```diff
@@ -0,0 +1,593 @@
+                .catch((error) => this.addError(error.response.data.err_msg));
+        },
+        setUsername() {
+            const Galaxy = getGalaxyInstance();
```
I don't think `setUsername` should be the responsibility of the sharing component. If you move this out into its own component I think that would look better.

mvdbeek commented on client/src/components/Sharing/Sharing.vue:
```diff
@@ -0,0 +1,593 @@
+                this.setSharing(this.actions.unpublish);
+            }
+        },
+        getModel() {
```
Maybe rename this to `getSharing`for consistency with `setSharing` ?

OlegZharkov committed:
fix typoe

thanks @mvdbeek

Co-authored-by: Marius van den Beek <m.vandenbeek@gmail.com>

OlegZharkov committed:
fix selenium

OlegZharkov committed:
rename getModel to getSharing

OlegZharkov committed:
refactor test

OlegZharkov committed:
add error message

OlegZharkov committed:
check that contains is not None

OlegZharkov committed:
wait for error to appear

OlegZharkov committed:
display proper error message

davelopez committed:
Explicitly return an empty response with 204

See https://github.com/tiangolo/fastapi/issues/2832

mvdbeek reviewed (approved):
That was a long journey, but it looks really good now. Thanks a lot @OlegZharkov @davelopez !
